src/test/java/ArrayIteratorTest.java

```diff
@@ -0,0 +1,176 @@
+import org.junit.jupiter.api.Test;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class ArrayIteratorTest {
+
+    // Note:
+    //      Size is the number of "meaningful" elements in the array
+    //      Capacity is the number of available indices in the array
+
+    @Test
+    void hasNext_capacityZeroArray_returnsFalse() {
+        Integer[] a = {};
+        Iterator<Integer> it = new ArrayIterator<>(a, 0);
+        assertFalse(it.hasNext());
+    }
+
+    @Test
+    void hasNext_fullCapacityOneArray_returnsTrue() {
+        Integer[] a = {10};
+        Iterator<Integer> it = new ArrayIterator<>(a, 1);
+        assertTrue(it.hasNext());
+    }
+
+    @Test
+    void hasNext_sizeZeroCapacityTenArray_returnsFalse() {
+        Integer[] a = {null, null, null, null, null, null, null, null, null, null};
+        Iterator<Integer> it = new ArrayIterator<>(a, 0);
+        assertFalse(it.hasNext());
+    }
+
+    @Test
+    void hasNext_sizeOneCapacityTenArray_returnsTrue() {
+        Integer[] a = {10, null, null, null, null, null, null, null, null, null};
+        Iterator<Integer> it = new ArrayIterator<>(a, 1);
+        assertTrue(it.hasNext());
+    }
+
+    @Test
+    void hasNext_sizeFiveCapacityTenArray_returnsTrue() {
+        Integer[] a = {10, 11, 12, 13, 14, null, null, null, null, null};
+        Iterator<Integer> it = new ArrayIterator<>(a, 5);
+        assertTrue(it.hasNext());
+    }
+
+    @Test
+    void hasNext_sizeOneCapacityTenArrayAfterHasNext_doesNotChangeState() {
+        Integer[] a = {10, null, null, null, null, null, null, null, null, null};
+        Iterator<Integer> it = new ArrayIterator<>(a, 1);
+        it.hasNext();
+        assertTrue(it.hasNext());
+        assertEquals(10, it.next());
+    }
+
+    @Test
+    void hasNext_sizeFiveCapacityTenArrayAfterFiveHasNext_doesNotChangeState() {
+        Integer[] a = {10, 11, 12, 13, 14, null, null, null, null, null};
+        Iterator<Integer> it = new ArrayIterator<>(a, 5);
+        it.hasNext();
+        it.hasNext();
+        it.hasNext();
+        it.hasNext();
+        it.hasNext();
+        assertTrue(it.hasNext());
+        assertEquals(10, it.next());
+    }
+
+    @Test
+    void hasNext_endOfSizeOneCapacityTenArray_returnsFalse() {
+        Integer[] a = {10, null, null, null, null, null, null, null, null, null};
+        Iterator<Integer> it = new ArrayIterator<>(a, 1);
+        it.next();
+        assertFalse(it.hasNext());
+    }
+
+    @Test
+    void hasNext_endOfSizeFiveCapacityTenArray_returnsFalse() {
+        Integer[] a = {10, 11, 12, 13, 14, null, null, null, null, null};
+        Iterator<Integer> it = new ArrayIterator<>(a, 5);
+        it.next();
+        it.next();
+        it.next();
+        it.next();
+        it.next();
```

Comment:
Minor, I'd prefer to see this written as a for loop (`for (int i=0; i < size; ++i) { it.next(); }`). It's easier imo to see the number of times the method is called that way. Applies to several tests.